dozer/cogs/filter.py

```diff
@@ -0,0 +1,266 @@
+import discord
+import re
+from discord.ext.commands import guild_only, has_permissions
+import datetime
+
+from ._utils import *
+from .. import db
+
+class Filter(Cog):
+
+	"""The filters need to be compiled each time they're run, but we don't want to compile every filter
+	Every time it's run, or all of them at once when the bot starts. So the first time that filter is run,
+	the compiled object is placed in here. This dict is actually a dict full of dicts, with each parent dict's key
+	being the guild ID for easy accessing.
+	"""
+	filter_dict = {}
+
+	"""Helper Functions"""
+
+	async def check_dm_filter(self, ctx):
+		with db.Session() as session:
+			results = session.query(WordFilterSetting).filter(WordFilterSetting.guild_id == ctx.guild.id,
+															  WordFilterSetting.setting_type == "dm")\
+				.one_or_none()
+
+			if results is None:
+				results = True
+			else:
+				results = results.value
+
+			if results is "1":
+				if ctx.author.dm_channel is None:
+					await ctx.author.create_dm()
+				return ctx.author.dm_channel
+			else:
+				return ctx.channel
+
+	def load_filters(self, guild_id):
+		with db.Session() as session:
+			results = session.query(WordFilter).filter(WordFilter.guild_id == guild_id,
+													   WordFilter.enabled == True).all()
+			self.filter_dict[guild_id] = {}
+			for filter in results:
+				self.filter_dict[guild_id][filter.id] = re.compile(filter.pattern, re.IGNORECASE)
+
+	async def check_filters(self, message):
+		if message.author.id == self.bot.user.id:
+			return
+		with db.Session() as session:
+			roles = session.query(WordFilterRoleWhitelist).filter(WordFilterRoleWhitelist.guild_id==message.guild.id,
+																  WordFilter.enabled == 1).all()
+		role_ids = []
+		for role in roles:
+			role_ids.append(role.role_id)
+		for user_role in message.author.roles:
+			if user_role.id in role_ids:
+				return  # I feel like there's a more efficient way of searching through the roles
+			# at least a more pythonic way at least
+		filters = {}
+		try:
+			filters = self.filter_dict[message.guild.id]
+		except KeyError:
+			self.load_filters(message.guild.id)
+			filters = self.filter_dict[message.guild.id]
+		for id, filter in filters.items():
+			if filter.search(message.content) is not None:
+				await message.delete()
+				await message.channel.send("{}, Banned word detected!".format(message.author.mention), delete_after=5.0)
+				time = datetime.datetime.now()
+				with db.Session() as session:
+					infraction = WordFilterInfraction(member_id=message.author.id, filter_id=id,
+													timestamp=time,
+													message=message.content)
+					session.add(infraction)
+
+	"""Event Handlers"""
+
+	async def on_message(self, message):
+		await self.check_filters(message)
+
+	async def on_message_edit(self, before, after):
+		await self.check_filters(after)
+
+	"""Commands"""
+
+	@group(invoke_without_command=True)
+	@guild_only()
+	async def filter(self, ctx, advanced = False):
```

Comment:
`advance` should be annotated as a `bool`, so `false`, `no`, etc will be converted to False.